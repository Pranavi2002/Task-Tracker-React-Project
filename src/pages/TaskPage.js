import React, { useState, useCallback } from "react";
import TaskForm from "../components/TaskForm";
import TaskList from "../components/TaskList";
import useTasks from "../hooks/useTasks";
import useTheme from "../hooks/useTheme";
import { Box, Typography, FormControl, InputLabel, Select, MenuItem } from "@mui/material";

export default function TaskPage() {
  const { tasks, addTask, deleteTask, editTask, toggleTask } = useTasks();
  const { mode } = useTheme(); // use context mode
  const [filter, setFilter] = useState("all");
  const [sortOrder, setSortOrder] = useState("none");

  const handleToggle = useCallback((id) => toggleTask(id), [toggleTask]);
  const handleDelete = useCallback((id) => deleteTask(id), [deleteTask]);
  const handleEdit = useCallback((id, updatedFields) => editTask(id, updatedFields), [editTask]);

  const filteredTasks = tasks.filter((t) => {
    if (filter === "completed") return t.completed;
    if (filter === "active") return !t.completed;
    return true;
  });

  const sortedTasks = [...filteredTasks].sort((a, b) => {
    if (!a.dueDate) return 1;
    if (!b.dueDate) return -1;
    if (sortOrder === "asc") return new Date(a.dueDate) - new Date(b.dueDate);
    if (sortOrder === "desc") return new Date(b.dueDate) - new Date(a.dueDate);
    return 0;
  });

  return (
    <Box
      sx={{
        p: 3,
        minHeight: "100vh",
        bgcolor: "background.default", // uses MUI theme colors
        color: "text.primary",
      }}
    >
      <Typography variant="h4" align="center" gutterBottom>
        Task Tracker
      </Typography>
      <TaskForm onAddTask={addTask} />

      <Box sx={{ display: "flex", justifyContent: "center", mt: 3, gap: 2 }}>
        <FormControl>
          <InputLabel>Filter</InputLabel>
          <Select value={filter} onChange={(e) => setFilter(e.target.value)}>
            <MenuItem value="all">All</MenuItem>
            <MenuItem value="active">Active</MenuItem>
            <MenuItem value="completed">Completed</MenuItem>
          </Select>
        </FormControl>

        <FormControl>
          <InputLabel>Sort by Due Date</InputLabel>
          <Select value={sortOrder} onChange={(e) => setSortOrder(e.target.value)}>
            <MenuItem value="none">None</MenuItem>
            <MenuItem value="asc">Earliest First</MenuItem>
            <MenuItem value="desc">Latest First</MenuItem>
          </Select>
        </FormControl>
      </Box>

      <TaskList tasks={sortedTasks} onToggle={handleToggle} onDelete={handleDelete} onEdit={handleEdit} />
    </Box>
  );
}