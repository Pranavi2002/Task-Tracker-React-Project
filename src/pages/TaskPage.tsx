import React, { useState, useCallback } from "react";
import TaskForm from "../components/TaskForm";
import TaskList from "../components/TaskList";
import useTheme from "../hooks/useTheme";
import useTasks, { Task } from "../hooks/useTasks";
import { motion } from "framer-motion";

const TaskPage: React.FC = () => {
  const { theme } = useTheme();
  const { tasks, addTask, deleteTask, editTask, toggleTask } = useTasks();

  const [filter, setFilter] = useState<"all" | "active" | "completed">("all");
  const [sortOrder, setSortOrder] = useState<"none" | "asc" | "desc">("none");

  const handleToggle = useCallback((id: number) => toggleTask(id), [toggleTask]);
  const handleDelete = useCallback((id: number) => deleteTask(id), [deleteTask]);
  const handleEdit = useCallback((id: number, updatedFields: Partial<Task>) => editTask(id, updatedFields), [editTask]);

  const filteredTasks = tasks.filter((t) => {
    if (filter === "completed") return t.completed;
    if (filter === "active") return !t.completed;
    return true;
  });

  const sortedTasks = [...filteredTasks].sort((a, b) => {
    if (!a.dueDate) return 1;
    if (!b.dueDate) return -1;
    if (sortOrder === "asc") return new Date(a.dueDate).getTime() - new Date(b.dueDate).getTime();
    if (sortOrder === "desc") return new Date(b.dueDate).getTime() - new Date(a.dueDate).getTime();
    return 0;
  });

  const containerStyle: React.CSSProperties = {
    padding: "20px",
    backgroundColor: theme === "light" ? "#f4f4f9" : "#222",
    color: theme === "light" ? "#222" : "#f4f4f9",
    minHeight: "100vh",
    fontFamily: "'Segoe UI', Tahoma, Geneva, Verdana, sans-serif",
  };

  const selectStyle: React.CSSProperties = {
    padding: "5px 8px",
    marginLeft: "10px",
    borderRadius: "5px",
    border: "1px solid #ccc",
  };

  return (
    <motion.div style={containerStyle}>
      <h1 style={{ textAlign: "center" }}>Task Tracker</h1>
      <TaskForm onAddTask={addTask} />

      <div style={{ marginTop: "20px", textAlign: "center" }}>
        <label>Filter: </label>
        <select style={selectStyle} value={filter} onChange={(e) => setFilter(e.target.value as "all" | "active" | "completed")}>
          <option value="all">All</option>
          <option value="active">Active</option>
          <option value="completed">Completed</option>
        </select>

        <label style={{ marginLeft: "20px" }}>Sort by Due Date: </label>
        <select style={selectStyle} value={sortOrder} onChange={(e) => setSortOrder(e.target.value as "none" | "asc" | "desc")}>
          <option value="none">None</option>
          <option value="asc">Earliest First</option>
          <option value="desc">Latest First</option>
        </select>
      </div>

      <TaskList tasks={sortedTasks} onToggle={handleToggle} onDelete={handleDelete} onEdit={handleEdit} />
    </motion.div>
  );
};

export default TaskPage;