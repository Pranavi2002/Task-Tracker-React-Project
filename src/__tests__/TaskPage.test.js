// src/__tests__/TaskPage.test.js
import React from "react";
import { render } from "@testing-library/react";
import TaskPage from "../pages/TaskPage";

// Mock hooks used in TaskPage
jest.mock("../hooks/useTheme", () => () => ({ theme: "light" }));
jest.mock("../hooks/useTasks", () => () => ({
  tasks: [
    { id: 1, text: "Task 1", completed: false, dueDate: "2025-08-24" },
    { id: 2, text: "Task 2", completed: true, dueDate: "" },
  ],
  addTask: jest.fn(),
  deleteTask: jest.fn(),
  editTask: jest.fn(),
  toggleTask: jest.fn(),
}));

describe("TaskPage Snapshot", () => {
  it("renders correctly", () => {
    const { asFragment } = render(<TaskPage />);
    expect(asFragment()).toMatchSnapshot(); // Save and compare page DOM structure
  });
});