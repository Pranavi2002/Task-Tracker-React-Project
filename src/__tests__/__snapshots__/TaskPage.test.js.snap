// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskPage Snapshot renders correctly 1`] = `
<DocumentFragment>
  <div
    style="padding: 20px; background-color: rgb(244, 244, 249); color: rgb(34, 34, 34); min-height: 100vh; font-family: 'Segoe UI', Tahoma, Geneva, Verdana, sans-serif;"
  >
    <h1
      style="text-align: center;"
    >
      Task Tracker
    </h1>
    <form
      style="margin-top: 20px;"
    >
      <input
        placeholder="Enter a task"
        style="padding: 8px; border-radius: 5px; border: 1px solid #ccc; margin-right: 10px;"
        type="text"
        value=""
      />
      <input
        style="padding: 8px; border-radius: 5px; border: 1px solid #ccc; margin-right: 10px;"
        type="date"
        value=""
      />
      <button
        style="padding: 6px 12px; border-radius: 5px; cursor: pointer; background-color: rgb(76, 175, 80); color: rgb(255, 255, 255); transition: 0.3s;"
        tabindex="0"
        type="submit"
      >
        Add
      </button>
    </form>
    <div
      style="margin-top: 20px; text-align: center;"
    >
      <label>
        Filter: 
      </label>
      <select
        style="padding: 5px 8px; margin-left: 10px; border-radius: 5px; border: 1px solid #ccc;"
      >
        <option
          value="all"
        >
          All
        </option>
        <option
          value="active"
        >
          Active
        </option>
        <option
          value="completed"
        >
          Completed
        </option>
      </select>
      <label
        style="margin-left: 20px;"
      >
        Sort by Due Date: 
      </label>
      <select
        style="padding: 5px 8px; margin-left: 10px; border-radius: 5px; border: 1px solid #ccc;"
      >
        <option
          value="none"
        >
          None
        </option>
        <option
          value="asc"
        >
          Earliest First
        </option>
        <option
          value="desc"
        >
          Latest First
        </option>
      </select>
    </div>
    <div
      style="position: relative; height: 400px; width: 100%; overflow: auto; will-change: transform; direction: ltr;"
    >
      <div
        style="height: 120px; width: 100%;"
      >
        <div
          style="position: absolute; left: 0px; top: 0px; height: 60px; width: 100%;"
        >
          <li
            style="margin-bottom: 10px; padding: 12px; border-radius: 8px; background-color: rgb(255, 255, 255); box-shadow: 0 1px 4px rgba(0,0,0,0.1); display: flex; justify-content: space-between; align-items: center; opacity: 0; transform: translateX(-50px);"
          >
            <span
              style="text-decoration: none; cursor: pointer; color: red;"
            >
              Task 1 (Due: 2025-08-24 ⚠️ Overdue)
            </span>
            <button
              style="padding: 5px 10px; margin-left: 5px; border-radius: 5px; cursor: pointer; background-color: rgb(135, 206, 235); color: rgb(255, 255, 255); transition: 0.3s;"
              tabindex="0"
            >
              Edit
            </button>
            <button
              style="padding: 5px 10px; margin-left: 5px; border-radius: 5px; cursor: pointer; background-color: rgb(211, 47, 47); color: rgb(255, 255, 255); transition: 0.3s;"
              tabindex="0"
            >
              ❌ Delete
            </button>
          </li>
        </div>
        <div
          style="position: absolute; left: 0px; top: 60px; height: 60px; width: 100%;"
        >
          <li
            style="margin-bottom: 10px; padding: 12px; border-radius: 8px; background-color: rgb(255, 255, 255); box-shadow: 0 1px 4px rgba(0,0,0,0.1); display: flex; justify-content: space-between; align-items: center; opacity: 0; transform: translateX(-50px);"
          >
            <span
              style="text-decoration: line-through; cursor: pointer;"
            >
              Task 2 
            </span>
            <button
              style="padding: 5px 10px; margin-left: 5px; border-radius: 5px; cursor: pointer; background-color: rgb(135, 206, 235); color: rgb(255, 255, 255); transition: 0.3s;"
              tabindex="0"
            >
              Edit
            </button>
            <button
              style="padding: 5px 10px; margin-left: 5px; border-radius: 5px; cursor: pointer; background-color: rgb(211, 47, 47); color: rgb(255, 255, 255); transition: 0.3s;"
              tabindex="0"
            >
              ❌ Delete
            </button>
          </li>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
